### PR TITLE
Improve TopoServer Performance and Efficiency For Keyspace Shards

### PR DESCRIPTION
**Pull request by mattlord**
## Description

There are various cases — e.g. when working with VReplication workflows — where we get all of the [serving] shards in a keyspace. To do this we were first getting a list of all shard names, then getting each shard record _**serially**_. This resulted in many topo server calls and when the topo server has high latency it could cause various commands to timeout and have some knock on effects on other things due to long running reads blocking other operations (especially e.g. with older etcd versions).

For example, if you had a keyspace with 128 shards then you would make 129 topo server calls when getting all [serving] shards:
  1. Get all shard names (1)
  2. Get each shard record serially (128)

With this PR you would go from 129 topo server calls in this example case down to 1 when the topo server supports key prefix scans (all but ZooKeeper do), and when the topo server does not (or the response message is beyond the max message size) then we fall back to the shard by shard method but do so _**concurrently**_. So either way, the total time taken to get all of the shards in a keyspace should improve dramatically.

## Related Issues

  - Fixes: https://github.com/vitessio/vitess/issues/15048

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not requiredk

### CODE REVIEW

**Comment by frouioui** [comment]
@mattlord, I think it would be great to add a benchmark test for this. Is it feasible to add an E2E benchmark test, maybe not with a 128 shards, but at least a significant amount?

File: go/vt/topo/keyspace.go
Comment (ajm188) [inline go/vt/topo/keyspace.go:207]: I _think_ you want

```go
if IsErrType(err, NoNode) || len(listResults) == 0
```

here Or does that situation never happen

File: go/vt/topo/keyspace.go
Comment (ajm188) [inline go/vt/topo/keyspace.go:315]: 👍

**Review by ajm188 — COMMENTED** [review]
this rules, just one question, but happy to approve whenever you're ready for a final pass

File: go/vt/topo/keyspace.go
Comment (frouioui) [inline go/vt/topo/keyspace.go:206]: I am confused why the `len(listResults) == 0` is required here. If we do not have an error but we have an empty result set, don't we want to go into the `else` block?

Comment (frouioui) [reply go/vt/topo/keyspace.go:206]: or generally, don't we always want to go into the `else` block if we don't have any error?

**Comment by mattlord** [comment]
> ## [Codecov](https://app.codecov.io/gh/vitessio/vitess/pull/15047?src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=vitessio) Report
> Attention: `15 lines` in your changes are missing coverage. Please review.

This is because the unit tests are cross package (dependency cycles prevented use of the same `topo` package). They will be counted once we're able to merge (waiting on next v4 release of the GitHub Action): https://github.com/vitessio/vitess/pull/14967

File: go/vt/topo/keyspace.go
Comment (deepthi) [inline go/vt/topo/keyspace.go:234]: nit: I believe it is idiomatic to handle the not-nil case first. In fact we should probably link to the standard golang style guide on our website for developers to use as guidance. It will be even better if we can somehow enforce golang style in CI so that it doesn't depend on reviewers' preferences.

File: go/vt/topo/keyspace.go
Comment (deepthi) [inline go/vt/topo/keyspace.go:46]: We are defaulting this to 32 for `GetTablets` so we now have two separate defaults depending on whether we are getting tablets or shards. The reasoning for 32 is covered by this comment. https://github.com/vitessio/vitess/pull/14693#discussion_r1417863256
I'm not completely opposed to changing the default for `GetTablets` to this value, but I'd prefer to change this 32 for the reasons listed in that comment.

**Review by deepthi — COMMENTED** [review]
Nice optimization.
As a follow up we can probably do another PR to replace [the call to `GetServingShards` in materializer.go](https://github.com/vitessio/vitess/blob/f1061648c0975a7add8d9dee744a7be95caffa9f/go/vt/wrangler/materializer.go#L395) with a special purpose func that returns only the first shard. Because that's all we ever use in that function.

File: go/vt/topo/keyspace.go
Comment (mattlord) [reply go/vt/topo/keyspace.go:46]: OK, I see that you were basing it on `fs.Int64Var(&topoReadConcurrency, "topo_read_concurrency", 32, "Concurrency of topo reads."), but you didn't reference the value. I'll see if I can use the variable somehow in both places.

File: go/vt/topo/keyspace.go
Comment (ajm188) [reply go/vt/topo/keyspace.go:234]: generally true, yes (exception is in a `switch-case`)

i like the idea of a CI check but it's probably not really feasible since a lot of the guidelines take the form of "prefer X" or "avoid Y" without outright banning many things. a lot (all?) of the mandates are covered by `gofmt` + `go vet`

**Review by ajm188 — APPROVED** [review]
approving, pending resolution of the concurrency default (which i don't have strong feelings on)

File: go/vt/schemamanager/tablet_executor.go
Comment (frouioui) [inline go/vt/schemamanager/tablet_executor.go:112]: Nit picking here but we are defining a default `&topo.FindAllShardsInKeyspaceOptions` in three different places, perhaps we should create a global default variable and use that here and in the two other places.

Comment (frouioui) [reply go/vt/schemamanager/tablet_executor.go:112]: In this file, in `go/vt/topo/test/shard.go` and in `go/vt/vtctl/workflow/utils.go`

File: go/vt/topo/keyspace.go
Comment (frouioui) [inline go/vt/topo/keyspace.go:212]: nit picking here too, it might read better if we extract the content of this entire if block. it seems like we only need to send in `listResults` as an argument and return `results, err`, something like:
```go
if err == nil {
	return handleResults(listResults)
}
```

**Review by frouioui — COMMENTED** [review]
It looks good to me! I will leave it up to @deepthi to give a second approval once her feedback has been addressed 😃

File: go/vt/topo/keyspace.go
Comment (mattlord) [reply go/vt/topo/keyspace.go]: I addressed both of your comments here: https://github.com/vitessio/vitess/pull/15047/commits/c7a27d0b60d25421cb8e6185a4b32b67f758cc83

File: go/vt/schemamanager/tablet_executor.go
Comment (mattlord) [reply go/vt/schemamanager/tablet_executor.go]: https://github.com/vitessio/vitess/pull/15047/commits/8a0638e4b5130d8dd279791f7db504aac31843f0

File: go/vt/topo/keyspace.go
Comment (mattlord) [reply go/vt/topo/keyspace.go]: Not sure it's more readable, but: https://github.com/vitessio/vitess/pull/15047/commits/3fb3e0421146a117fc2979b148265ed365adf9df

File: go/vt/topo/keyspace.go
Comment (deepthi) [inline go/vt/topo/keyspace.go:47]: I like this! Instead of hard-coding the default for all but healthcheck, we now allow this to be customized for any binary that imports the topo package.

File: go/vt/topo/keyspace.go
Comment (deepthi) [inline go/vt/topo/keyspace.go:44]: Ugh, I just realized the flag value was always int64, and there's no good reason for that. Should we change this to int? Is there any reason not to?

**Review by deepthi — APPROVED** [review]
It will be nice to address the int64 vs int issue. Rest LGTM.

File: go/vt/topo/keyspace.go
Comment (ajm188) [reply go/vt/topo/keyspace.go:44]: No reason not to. `int` is `int64` on 64-bit platforms, and for 32-bit they can't really be using bigger numbers anyway 🤷

Comment (mattlord) [reply go/vt/topo/keyspace.go:44]: 32 bit machines support 64 bit integers.

Comment (mattlord) [reply go/vt/topo/keyspace.go:44]: You can use int64 on 32 bit machines -- two 32 bit pieces are used. We (the royal we) were using long long / int64 when needed when most machines were still 32 bit architectures.

Comment (mattlord) [reply go/vt/topo/keyspace.go:44]: I'm fine moving this to int though. I initially started there but then preferred not to change the existing flag.

Comment (mattlord) [reply go/vt/topo/keyspace.go]: Changed it here (and now using it in another location where we should be): https://github.com/vitessio/vitess/pull/15047/commits/99a7f86fbee1f6e4fa226984e93d20f80ee978ec

**Review by ajm188 — APPROVED** [review]

File: go/vt/discovery/healthcheck.go
Comment (deepthi) [inline go/vt/discovery/healthcheck.go:111]: Good catch. This was unused 🤦

**Review by deepthi — APPROVED** [review]

**Comment by deepthi** [comment]
Failures in the Code coverage workflow are being caused by flaky tests which will be fixed separately.

**Pull request merged**
